### PR TITLE
Experimental setting and adapter launcher for VSCode based debugger 

### PR DESCRIPTION
**Pull request by karthiknadig**
If the experimental flag is enabled and the debugger is not available this is the error you will see:
image

### CODE REVIEW

File: Python/Product/Debugger/DebugAdapterLauncher.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using Microsoft.PythonTools.DkmDebugger;
```
Comment (int19h) [inline Python/Product/Debugger/DebugAdapterLauncher.cs]: What's coming from this namespace? It shouldn't be needed for pure Python debugging.

Also, can you sort the usings?

File: Python/Product/PythonTools/PythonTools/Debugger/DebugLaunchHelper.cs [hunk 2]
```diff
@@ -84,40 +85,82 @@ public static class DebugLaunchHelper {
             return SubstitutionPattern.Replace(
                 str,
                 m => {
```
Comment (int19h) [inline Python/Product/PythonTools/PythonTools/Debugger/DebugLaunchHelper.cs]: No need for braces or `return` now, either.

File: Python/Product/PythonTools/PythonTools/Debugger/DebugLaunchHelper.cs [hunk 3]
```diff
@@ -84,40 +85,82 @@ public static class DebugLaunchHelper {
...
+            var pyService = provider.GetPythonToolsService();
+            return string.Join(";", GetGlobalDebuggerOptions(pyService)
+                .Concat(GetLaunchConfigurationOptions(config))
+                .Where(s => !string.IsNullOrEmpty(s))
+                .Select(s => s.Replace(";", ";;"))
+            );
+        }
+
+        private static string GetLaunchJsonForVsCodeDebugAdapter(IServiceProvider provider, LaunchConfiguration config) {
+            bool nativeDebug = config.GetLaunchOption(PythonConstants.EnableNativeCodeDebugging).IsTrue();
```
Comment (int19h) [inline Python/Product/PythonTools/PythonTools/Debugger/DebugLaunchHelper.cs]: This should never be true on this code path. Basically, if native debugging is enabled, we should be using the old engine (which in that case just bails out and lets Concord handle it).

File: Python/Product/PythonTools/SharedProject/Debugger/ProvideDebugAdapterAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+             * This appears in VS in several places.  For example:
+             *     -The "Select Code Type" dialog for choosing which debugger to attach to a process(if Attach is supported)
+             *     -The "Debugging" column of the "Processes" tool window
+             */
+            engineKey.SetValue("Name", _name);
+
+            /*
+             * Path to the debug adapter executable
+             */
+            engineKey.SetValue("Adapter", @"$PackageFolder$\DebugAdapter.exe");
```
Comment (int19h) [inline Python/Product/PythonTools/SharedProject/Debugger/ProvideDebugAdapterAttribute.cs]: Do we actually need this, given that we'll have a custom `IAdapterLauncher` that provides the streams directly, without another process?

File: Python/Product/PythonTools/SharedProject/Debugger/ProvideDebugAdapterAttribute.cs [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+             * This appears in (e.g.) the "Language" column of the Stack Trace tool window.
+             */
+            engineKey.SetValue("Language", _languageName);
+            engineKey.SetValue("LanguageId", _languageId);
+
+            /* 
+             * Adapter launcher registration 
+             */
+            var adapterKey = context.CreateKey($"CLSID\\{_adapterLauncherCLSID}");
+            adapterKey.SetValue("Assembly", "Microsoft.PythonTools.Debugger");
```
Comment (int19h) [inline Python/Product/PythonTools/SharedProject/Debugger/ProvideDebugAdapterAttribute.cs]: Might be a good idea to use `typeof(...).FullName` here, to make this refactoring-proof.

File: Python/Product/Debugger/DebugAdapterProcess.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using Microsoft.VisualStudio.Debugger.Interop.VSCodeDebuggerHost;
+
+namespace Microsoft.PythonTools {
```
Comment (zooba) [inline Python/Product/Debugger/DebugAdapterProcess.cs]: Can these go in the Debugger namespace?

File: Python/Product/Debugger/Debugger.csproj [hunk 7]
```diff
@@ -46,6 +46,11 @@
     <Reference Include="envdte80, Version=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" />
     <Reference Include="Microsoft.VisualStudio.ComponentModelHost, Version=$(VSTarget).0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL" />
     <Reference Include="Microsoft.VisualStudio.Debugger.Engine, Version=1.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL" />
+    <Reference Include="Microsoft.VisualStudio.Debugger.Interop.VSCodeDebuggerHost, Version=1.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL" />
+    <Reference Include="Microsoft.VisualStudio.Debugger.InteropA, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
+      <HintPath>..\..\packages\Microsoft.VisualStudio.Debugger.InteropA.9.0.21022\lib\Microsoft.VisualStudio.Debugger.InteropA.dll</HintPath>
+      <Private>True</Private>
```
Comment (zooba) [inline Python/Product/Debugger/Debugger.csproj]: Neither of these properties should be needed. We default all references to non-private (since they mostly are), interop types should be embedded, and all the nuget packages get added to the assembly search path automatically.

Setting any of them in the project file is likely to break that (and I know this was what VS added - just clarifying why we need to override them).

File: Python/Product/Debugger/Debugger.csproj [hunk 8]
```diff
@@ -239,6 +245,7 @@
       <SubType>Designer</SubType>
     </None>
     <Content Include="DkmDebugger\PythonDkm.natvis" />
+    <None Include="packages.config" />
```
Comment (zooba) [inline Python/Product/Debugger/Debugger.csproj]: Don't add nuget dependencies directly to the project. Add them to the `Build\15.0\packages.config` file and run `PreBuild.ps1` again.

File: Python/Product/PythonTools/PythonTools.csproj [hunk 9]
```diff
@@ -692,6 +692,7 @@
     <Compile Include="Properties\AssemblyInfo.cs" />
     <Compile Include="PkgCmdID.cs" />
     <Compile Include="PythonTools\WebProject\PythonWebPropertyPage.cs" />
+    <Compile Include="SharedProject\Debugger\ProvideDebugAdapterAttribute.cs" />
```
Comment (zooba) [inline Python/Product/PythonTools/PythonTools.csproj]: Don't add files to the SharedProject folder - it isn't real (it's actually under `Common` rather than `Python`, and all its files are linked in).

This one can go under `PythonTools\Debugger`.

File: Python/Product/PythonTools/PythonTools/Debugger/DebugLaunchHelper.cs [hunk 10]
```diff
@@ -83,41 +84,79 @@ public static class DebugLaunchHelper {
...
+            JObject jsonObj = new JObject {
+                ["exe"] = config.GetInterpreterPath(),
+                ["curDir"] = string.IsNullOrEmpty(config.WorkingDirectory) ? PathUtils.GetParent(config.ScriptName) : config.WorkingDirectory,
+                ["remoteMachine"] = "",
+                ["args"] = GetArgs(config),
+                ["options"] = GetOptions(provider, config),
+                ["env"] = envArray
+            };
+            
+            // Note: these are optional, but sepcial. These override the pkgdef version of the adapter settings
```
Comment (zooba) [inline Python/Product/PythonTools/PythonTools/Debugger/DebugLaunchHelper.cs]: "special"

File: Python/Product/PythonTools/PythonTools/Options/ExperimentalOptions.cs [hunk 11]
```diff
@@ -61,5 +64,11 @@ public sealed class ExperimentalOptions {
         /// True to use conda for package management when available.
         /// </summary>
         public bool UseCondaPackageManager { get; set; }
+
+        /// <summary>
+        /// True to use new Ptvs debugger built to run on VS Code Debug Adapter Host
```
Comment (zooba) [inline Python/Product/PythonTools/PythonTools/Options/ExperimentalOptions.cs]: Always "PTVS"

Comment (zooba) [reply Python/Product/PythonTools/PythonTools/Options/ExperimentalOptions.cs]: And rarely needs to be named within our own project - this could just read "Use a debugger built to run on VS Code Debug Adapter Host"

File: Python/Product/PythonTools/PythonTools/Options/PythonExperimentalOptionsControl.Designer.cs [hunk 12]
```diff
@@ -26,9 +26,10 @@ partial class PythonExperimentalOptionsControl {
             System.ComponentModel.ComponentResourceManager resources = new System.ComponentModel.ComponentResourceManager(typeof(PythonExperimentalOptionsControl));
             this._noDatabaseFactory = new System.Windows.Forms.CheckBox();
             this.tableLayoutPanel2 = new System.Windows.Forms.TableLayoutPanel();
-            this._mustRestartLabel = new System.Windows.Forms.Label();
```
Comment (zooba) [inline Python/Product/PythonTools/PythonTools/Options/PythonExperimentalOptionsControl.Designer.cs]: Double-check that the tab order has not changed in this dialog.

File: Python/Product/VSInterpreters/ExperimentalOptions.cs [hunk 13]
```diff
@@ -23,13 +23,16 @@ public static class ExperimentalOptions {
         internal const string NoDatabaseFactoryKey = "NoDatabaseFactory";
         internal const string AutoDetectCondaEnvironmentsKey = "AutoDetectCondaEnvironments";
         internal const string UseCondaPackageManagerKey = "UseCondaPackageManager";
+        internal const string UseVsCodeDebuggerKey = "UseVsCodeDebugger";
         internal static readonly Lazy<bool> _noDatabaseFactory = new Lazy<bool>(GetNoDatabaseFactory);
         internal static readonly Lazy<bool> _autoDetectCondaEnvironments = new Lazy<bool>(GetAutoDetectCondaEnvironments);
         internal static readonly Lazy<bool> _useCondaPackageManager = new Lazy<bool>(GetUseCondaPackageManager);
+        internal static readonly Lazy<bool> _useVsCodeDebugger = new Lazy<bool>(GetUseNewPtvsDebugger);
```
Comment (zooba) [inline Python/Product/VSInterpreters/ExperimentalOptions.cs]: GetUseVsCodeDebugger

File: Python/Product/VSInterpreters/ExperimentalOptions.cs [hunk 14]
```diff
@@ -23,13 +23,16 @@ public static class ExperimentalOptions {
...
 
         public static bool GetNoDatabaseFactory() => GetBooleanFlag(NoDatabaseFactoryKey, defaultVal: true);
         public static bool GetAutoDetectCondaEnvironments() => GetBooleanFlag(AutoDetectCondaEnvironmentsKey, defaultVal: false);
         public static bool GetUseCondaPackageManager() => GetBooleanFlag(UseCondaPackageManagerKey, defaultVal: false);
+        public static bool GetUseNewPtvsDebugger() => GetBooleanFlag(UseVsCodeDebuggerKey, defaultVal: false);
```
Comment (zooba) [inline Python/Product/VSInterpreters/ExperimentalOptions.cs]: GetUseVsCodeDebugger

File: Python/Product/Debugger/Debugger.csproj (hunk 7 above)
Comment (int19h) [reply Python/Product/Debugger/Debugger.csproj]: We don't embed any interop assemblies in this project (note the references below). I don't think that's intentional, though.

Comment (zooba) [reply Python/Product/Debugger/Debugger.csproj]: Yeah, you're right. Embedding interop types isn't actually the default.

General point is, use the default settings until they fail.

File: Python/Product/PythonTools/PythonTools/Intellisense/ExtractMethodSuggestedActionSourceProvider.cs [hunk 15]
```diff
@@ -82,7 +82,9 @@ class ExtractMethodSuggestedActionSource : ISuggestedActionsSource {
                 if (MethodExtractor.CanExtract(_view) == true) {
                     return new SuggestedActionSet[] {
                     new SuggestedActionSet(
+                        requestedActionCategories.First(),
```
Comment (zooba) [inline Python/Product/PythonTools/PythonTools/Intellisense/ExtractMethodSuggestedActionSourceProvider.cs]: Unrelated change? Should safely handle `InvalidOperationException` if you're going to use `First()`.

Comment (karthiknadig) [reply Python/Product/PythonTools/PythonTools/Intellisense/ExtractMethodSuggestedActionSourceProvider.cs]: I am going to undo this one. I was not able to build because VS kept complaining that the version that is used was Obsolete.
